Update backward formula for torch.dot and add backward definition for torch.vdot

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45074 Update backward formula for torch.dot and add backward definition for torch.vdot**

TODO: Add R -> C tests in https://github.com/pytorch/pytorch/pull/44744 (blocked on some JIT changes)

Differential Revision: [D23975361](https://our.internmc.facebook.com/intern/diff/D23975361)

anjali411 committed:
Update backward formula for torch.dot and add backward definition for torch.vdot

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -404,12 +404,12 @@
+  self: correct_dtype_gradients(self.scalar_type(), grad * tensor.conj())
+  tensor: correct_dtype_gradients(tensor.scalar_type(), grad * self.conj())
 
 - name: vdot(Tensor self, Tensor other) -> Tensor
```
What is vdot actually doing? It is not documented anywhere....

anjali411 replied:
https://pytorch.org/docs/master/generated/torch.vdot.html#torch.vdot

albanD replied:
Ho it's only on master! Thanks for the link.

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -945,6 +945,7 @@ def method_tests():
         ('addr', (S, M), ((S,), (M,)), 'coef', (), (), (), ident, {'beta': 0.2, 'alpha': 0.6}),
         ('addr', (), ((S,), (M,)), 'broadcast_lhs_coef', (), (), (), ident, {'beta': 0.2, 'alpha': 0.6}),
         ('dot', (L,), ((L,),), '', (True,)),
+        ('vdot', (L,), ((L,),), '', (True,)),
```
This should be (False,) at the end (to turn off JIT autodiff testing)

zou3519 replied:
Alternatively, `('vdot', (L,), ((L,),),` does the trick

anjali411 replied:
cc. @eellison `torch.vdot` is not supported by JIT autodiff right now.

zou3519 replied:
The question here is: Is it OK that autodiff doesn't support vdot? It's a new operator we added recently. Also, does the JIT still use the old autodiff pass?

eellison replied:
It should be fine if it's not supported by JIT autodiff - having an autodiff is an optimization which allows fusion to occur in training.  

Yea it still uses the old autodiff pass. The way to define a backwards is in the `symbolic_script.cpp` file. Currently, there is really only autodiff coverage for pointwise ops because those are the ops that we codegen fusion for.

anjali411 committed:
Update on "Update backward formula for torch.dot and add backward definition for torch.vdot"


TODO: Add R -> C tests in https://github.com/pytorch/pytorch/pull/44744 (blocked on some JIT changes)


[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -87,6 +87,14 @@ static Tensor wrapped_scalar_tensor(Scalar scalar) {
   return tensor;
 }
 
+Tensor correct_dtype_gradients(ScalarType self_st, Tensor gradient_result) {
+  if (!at::isComplexType(self_st) && gradient_result.is_complex()) {
+    // R -> C
+    return at::real(gradient_result);
+  }
+  return gradient_result;
+}
```
"correct_dtype_gradients" seems like a very generic name (e.g. it could be mistaken for a function that handles float -> double dtype conversion). Also, it looks like this function will be used quite a lot in autograd formulas.

Tossing some quick ideas out there:
- "handle_r_to_c", "handle_real_to_complex"
- "maybe_real_part"

ezyang replied:
It seems to me that `at::real` should noop on real tensors, similar to how `at::conj` is noop on real tensors too.

anjali411 replied:
@zou3519 makes sense.`handle_real_to_complex` should be clearer, I think.

@ezyang yeah I agree. The reason we disabled `at::real` for non-complex tensors before was because it would be weird to have real return a view for non-complex tensors and `at::imag` return a new tensor populated with zeros (which it was before). We can certainly only enable `at::real` for non-complex tensors if we want though.

ezyang replied:
Sure, but now that `at::conj` is a no-op for real tensors, I think we should probably be OK with making `at::real` do the same as well. I don't care... too much about `imag`, I don't think it shows up in situations like this.

ezyang commented:
Besides Richard's comment, rest of the PR looks reasonable.

anjali411 committed:
Update on "Update backward formula for torch.dot and add backward definition for torch.vdot"


TODO: Add R -> C tests in https://github.com/pytorch/pytorch/pull/44744 (blocked on some JIT changes)


[ghstack-poisoned]

zou3519 reviewed (approved):
LGTM after the name change. The discussion on at::real being a no-op for real tensors seems orthogonal so I am approving to unblock.

anjali411 committed:
Update on "Update backward formula for torch.dot and add backward definition for torch.vdot"


TODO: Add R -> C tests in https://github.com/pytorch/pytorch/pull/44744 (blocked on some JIT changes)

Differential Revision: [D23975361](https://our.internmc.facebook.com/intern/diff/D23975361)

[ghstack-poisoned]
